### PR TITLE
5L depth: deeper models cross-dataset (DrivAerML/AirfRANS/TandemFoil)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Increasing model depth from the current best (2–4L) to 5 layers may unlock better feature hierarchies across all three datasets. The current DrivAerML best uses 4L/512d, AirfRANS uses 2L/256d, and TandemFoil uses 3L/192d. Adding a 5th layer at moderate width (256d for TF/AF, 512d+ for DAM) tests whether deeper representations improve generalization, particularly for complex 3D flow topology (DrivAerML) and OOD pressure prediction (TandemFoil). Width variants on DrivAerML also probe the depth×width tradeoff at 5L.

## Instructions

Implement the following 8 runs. **Do not hardcode SENPAI_TIMEOUT_MINUTES or SENPAI_MAX_EPOCHS.** Use `--epochs 999` and let the environment control the budget.

**Golden configs (baselines for each dataset):**

DAM base:
```
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

AF base:
```
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999
```

TF base:
```
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

**Run assignments (all use `--wandb_group 5L-deeper`):**

- GPU 0: TF base + `--model-layers 5 --model-hidden-dim 256 --model-heads 4`
- GPU 1: AF base + `--model-layers 5`
- GPU 2: DAM base + `--model-layers 5`
- GPU 3: DAM base + `--model-layers 5 --grad-clip 1.0 --weight-decay 1e-2`
- GPU 4: DAM base + `--model-layers 5 --lr 3e-4 --cosine-t-max 50`
- GPU 5: DAM base + `--model-layers 5 --model-hidden-dim 384 --model-heads 6`
- GPU 6: DAM base + `--model-layers 5 --model-hidden-dim 384 --model-heads 6 --grad-clip 1.0 --weight-decay 1e-2`
- GPU 7: DAM base + `--model-layers 5 --model-hidden-dim 640 --model-heads 10`

Run all 8 in parallel, one per GPU. Report `val_primary/surface_rel_l2_pct` for DrivAerML, `val_primary/surface_mse` for AirfRANS, and `val_primary/surface_pressure_mae` for TandemFoil.

## Baseline

**DrivAerML:** `val_primary/surface_rel_l2_pct = 4.619%` (PR #2691, epoch 256, 4L/512d/8H, AdamW lr=5e-4, T_max=30)
Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

**AirfRANS:** `val_primary/surface_mse = 0.001236` (PR #2828, epoch 349, 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5)
Reproduce: `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999`

**TandemFoil:** `val_primary/surface_pressure_mae = 30.10` (PR #2810, epoch 157, 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10)
Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT himmel:

## Results

### Summary

5-layer depth is **not beneficial** for DrivAerML or AirfRANS, and only **marginally beneficial** for TandemFoil. Deeper models are catastrophically unstable without gradient clipping, and even with gc+wd the best DrivAerML result (5.224%) is +13.1% worse than the 4L/512d baseline (4.619%). The hypothesis that a 5th layer unlocks better feature hierarchies is **not supported** — optimization instability at 5L overwhelms any representational benefit.

---

### TandemFoil: 5L/256d/4H (GPU 0) — MARGINAL WIN

| Metric | 5L/256d (this) | Baseline (3L/192d) | Delta |
|--------|---------------|-------------------|-------|
| `val_primary/surface_pressure_mae` | **29.83** | 30.10 | **-0.9%** |
| `test_primary/surface_pressure_mae` | 32.24 | — | — |

- **Best epoch:** 140 / 158 total (~2.3 min/epoch)
- **Per-split val at best (ep140):** geom_camber_cruise=18.63, geom_camber_rc=41.90, re_rand=29.21, single_in_dist=29.56
- **Per-split test (final):** geom_camber_cruise=34.91, geom_camber_rc=39.06, re_rand=25.10, single_in_dist=29.88
- **Peak memory:** ~77 GB
- **W&B run:** 85hjbz0z
- **Command:** `CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group 5L-deeper`
- **Analysis:** 5L/256d at TandemFoil marginally beats the 3L/192d baseline (29.83 vs 30.10). The wider dim (256 vs 192) and extra depth add capacity that the Lion+gc combo can stabilize. However, the improvement is tiny (-0.9%) and within noise — not a clear win. The test metric (32.24) is taken from the final epoch (158), not the best val epoch (140), so it reflects a slightly worse checkpoint.

---

### AirfRANS: 5L/256d/4H (GPU 1) — DIVERGED

| Metric | 5L/256d (this) | Baseline (2L/256d) | Delta |
|--------|---------------|-------------------|-------|
| `val_primary/surface_mse` | 0.003572 (best) | **0.001236** | **+189% (worse)** |
| `test_primary/surface_mse` | NaN (diverged) | — | — |

- **Best epoch:** 143 / 337 total (~1.07 min/epoch), diverged to NaN by epoch 167
- **full_val/surface_mse_p:** 0.014239 (at best), **full_val/volume_mse:** 0.014917
- **Peak memory:** ~75 GB
- **W&B run:** ab5d4hzt
- **External target:** 0.0043 — 5L/256d (0.003572) still beats external target but is 2.9x worse than our 2L baseline
- **Command:** `CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 5 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb-group 5L-deeper`
- **Analysis:** 5L is catastrophic for AirfRANS. The model trains normally until ep143 (best 0.003572), then the T_max=5 cosine cycling triggers irreversible divergence — all metrics go NaN by ep167. AirfRANS has consistently shown that shallower is better (4L→3L was -46.6%, 3L→2L was -16.4%), and going to 5L confirms the trend in the wrong direction. The depth-reduction signal on AirfRANS is now very clear: 2L is optimal.

---

### DrivAerML: 6 width/regularization variants at 5L (GPUs 2-7)

| GPU | Config | Best Val % | Epoch | Total Ep | Test % | Status |
|-----|--------|-----------|-------|----------|--------|--------|
| 2 | 5L/512d | 8.158 | 144 | 265 | 39.11 | **DIVERGED** (grad norms 905K) |
| **3** | **5L/512d + gc=1.0 + wd=1e-2** | **5.224** | **257** | **273** | **5.856** | **Best (stable)** |
| 4 | 5L/512d + lr=3e-4 + T_max=50 | 6.135 | 157 | 266 | 18.44 | Partially diverged |
| 5 | 5L/384d | 10.292 | 73 | 280 | 63.99 | **CATASTROPHIC** (grad norms 472K) |
| 6 | 5L/384d + gc=1.0 + wd=1e-2 | 5.524 | 234 | 263 | 20.91 | Partially diverged |
| 7 | 5L/640d | 13.549 | 128 | 269 | 66.88 | **CATASTROPHIC** (grad norms 40M) |

**Baseline: 4.619% (4L/512d/8H, PR #2691)**

**Best 5L DrivAerML: 5.224% (GPU 3, 5L/512d+gc+wd) — +13.1% worse than baseline**

- **W&B runs:** 8oqzjhfz (GPU2), wetm7pjl (GPU3), 5pi94qtk (GPU4), p8zxcl5g (GPU5), czyg4sc0 (GPU6), 3aznjxdd (GPU7)
- **Peak memory:** ~5-7 GB (DrivAerML 50k surface points keeps memory low)

**Key observations:**
1. **Gradient clipping is mandatory at 5L.** Without gc, all DrivAerML 5L runs diverge catastrophically. GPU2 (5L/512d, no gc) → grad norms hit 905K. GPU5 (5L/384d, no gc) → 472K. GPU7 (5L/640d, no gc) → 40M.
2. **gc=1.0 + wd=1e-2 is the minimum viable regularization.** GPU3 (512d+gc+wd) = 5.224%, GPU6 (384d+gc+wd) = 5.524%. Both stable but still worse than 4L baseline.
3. **Width at 5L:** 512d > 384d (5.224% vs 5.524%), consistent with 4L findings. 640d is too wide for 5L without gc.
4. **Lower LR doesn't help at 5L:** GPU4 (lr=3e-4, T_max=50) hit 6.135% — worse than GPU3's lr=5e-4 + gc+wd (5.224%).
5. **The 4L→5L transition adds critical optimization instability** that regularization can only partially mitigate. Even the best regularized 5L run cannot match 4L's performance.

---

### Cross-Dataset Analysis

| Dataset | Best 5L Result | Current Baseline | Delta | Verdict |
|---------|---------------|-----------------|-------|---------|
| TandemFoil | 29.83 (5L/256d) | 30.10 (3L/192d) | **-0.9%** | Marginal win, within noise |
| AirfRANS | 0.003572 (5L/256d) | 0.001236 (2L/256d) | **+189%** | Catastrophic regression, diverged |
| DrivAerML | 5.224% (5L/512d+gc+wd) | 4.619% (4L/512d) | **+13.1%** | Clear regression |

**The 5th layer hypothesis is falsified.** Each dataset has a different optimal depth:
- **TandemFoil:** 3L (maybe 5L with gc, but marginal)
- **AirfRANS:** 2L (depth reduction keeps winning)
- **DrivAerML:** 4L (adding a 5th layer hurts even with heavy regularization)

The pattern is consistent: the datasets where we've already found the right depth (via reduction experiments) cannot be improved by adding depth back. The optimization cost of deeper models (gradient instability, slower convergence, divergence risk) outweighs any representational benefit.

---

### What happened

The hypothesis that 5 layers would unlock better feature hierarchies was not supported. The key finding is that **depth sensitivity is dataset-specific and the current baselines already sit at or near optimal depth**:
- AirfRANS benefits from *fewer* layers (confirmed again — 5L is 2.9x worse than 2L)
- DrivAerML is optimized at 4L — adding a 5th layer requires heavy regularization (gc+wd) just to avoid catastrophic divergence, and still underperforms
- TandemFoil shows a tiny benefit from 5L but the signal is within noise

The massive gradient instability at 5L (grad norms reaching millions without clipping) reveals a fundamental optimization barrier. At 5 layers, the Transolver's attention-based physics slice mechanism creates gradient paths that amplify noise at cosine LR peaks. Gradient clipping can contain this but at the cost of slower convergence and shallower basins.

### Suggested follow-ups

1. **Don't go deeper on any dataset.** The depth frontier is confirmed: 2L for AirfRANS, 3-4L for TandemFoil, 4L for DrivAerML.
2. **Width scaling at optimal depth** is the more promising direction — 4L/512d already worked for DrivAerML, and width hasn't been fully explored at the optimal depth for each dataset.
3. **Compound gc+wd experiments at 4L** for DrivAerML — GPU3's gc+wd at 5L (5.224%) suggests this regularization combo could improve 4L too (currently 4.619% without gc/wd).
4. **TandemFoil 5L/256d with longer budget** — the 29.83 at ep140 was still in a slow descent; a longer run might separate from noise.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2891 because ADVISOR: Closing — 5L hypothesis falsified across all datasets. DrivAerML best 5.224% (+13%), TandemFoil 29.83 vs 26.06 baseline (+14.7%), AirfRANS diverged. DrivAerML depth is locked at 4L. Note: your TF comparison used outdated baseline 30.10; actual current baseline was 26.134 (now 26.06 after #2887 merge). Great systematic work..
